### ISSUE TITLE
variance in NCC loss different from the formula 6. in the paper

### ISSUE THREAD

**Issue by flealq** [voxelmorph/voxelmorph#226]
as ./torch/losses.py, line 58~63, 

        u_I = I_sum / win_size
        u_J = J_sum / win_size

        cross = IJ_sum - u_J * I_sum - u_I * J_sum + u_I * u_J * win_size
        I_var = I2_sum - 2 * u_I * I_sum + u_I * u_I * win_size
        J_var = J2_sum - 2 * u_J * J_sum + u_J * u_J * win_size

this is rather odd. 

take  (u_I = I_sum / win_size) into (I_var = I2_sum - 2 * u_I * I_sum + u_I * u_I * win_size)
then I_var = I2_sum - 2 *  I_sum / win_size * I_sum +  I_sum / win_size *  I_sum / win_size * win_size
then **I_var = I2_sum -  I_sum  * I_sum / win_size**

that is, (a-b)<sup>2</sup> = a<sup>2</sup> - b<sup>2</sup> ?

**Comment by flealq** [comment]
@adalca thank you for your reply both in #227 and this one. sorry for tagging wrong person... i'm not familier with github...

for your reply in #227 and for @balakg :
> 1. Can you please just continue asking on the previous issue #226 ? Even if it's closed, you can continue the discussion :) This allows us to track issues more easily.

i try to do so... and i closed #227. hope i'm doing the right thing. 

> 2. Oh, we agree that the code could be simplified -- in fact, in the keras implementation we have a comment with a `#TODO` to simplifythe expression. So I agree on computational complexity. But we've found to not matter much in practice that's why we haven't rushed to it. But I don't think it's wrong :) I don't quite know why you say "(1) is definitely not the paper means to be" -- if you expand the top of (6) I do think that's exactly what it comes out to, unless I am misunderstanding your point. Again, I'll tag @balakg  who might comment.

yes, my point is, expand (6) in the paper not equal to the code's implementation. simplify is not needed, i just to make it a more clear one.

> 3. Hmm, I don't believe so. Maybe you can write out the math and show us why you are confused? If you expand the sum, and then imagine doing a conv only for the terms that have a $p_i$ dependency (not those that have `p`),
> 4. and 4. I apprecaite you sharing the code, but unfortunately we don't have time to go over and debug code right now. I will note that errors of 1e-8 might be within the precision of various computations involving tensorflow's float16 (not sure)

ah... my word lead to missunderstanding. **script in 4 can run without debug.** "my code do not work" means in my application (which is about image decomposition but not registration), ncc not work so well.

btw, after one day training, my decomposition code like "def calc_cross_my(I, J)" in 4 got quite good result.
so if current code is right, i think i must misunderstand the paper. so maybe i can take it as my innovation, haha.

**Comment by adalca** [comment]
@flealq We'll take a look, I can't quite see the issue but worth checking into. Unfortunately my time coming up is quite limited, but I will try. Also tagging @balakg who understands the code more than I do.

**Comment by kvttt** [comment]
Hi, @flealq I was looking at your question earlier since I am also trying to understand the NCC loss used in the paper and I came across your questions here. I did a little bit of math here: 

![IMG_0105](https://user-images.githubusercontent.com/99367900/172707289-f04824f3-db30-418e-9ca3-b961d11b2292.PNG)

And I can confirm that your statement that 
> then I_var = I2_sum - I_sum * I_sum / win_size

is correct. But I believe your observation that (a-b)^2 = a^2 - b^2 is incorrect because the squared term is actually multiplied by win_size instead of the squared term by itself.
